### PR TITLE
Properly ignore joins on Public attributes

### PR DESCRIPTION
**Pull request by kmshin1397**
This PR is essentially a follow-up to Danny's PR #221. It reinforces the reasoning behind the changes there.  Most of these changes are just from black formatting, the relevant changes are lines 480:482 and 587:591 (I'll highlight these in a comment).

Basically, in #221, the "ignore and continue if combining a `Public` attribute" logic was added after the `class_attribute` for said attribute was already joined into the query ([here](https://github.com/cesium-ml/baselayer/blob/be52c9e6c8a74096e2ff154d0067f152e0c74aa8/app/models.py#L466)) so queries were still joining those in. The changes in lines 480:482 addresses this, and is responsible for the sample effects I'll describe below. 

This came up in the query for accessible Classifications on a particular Obj, generated [here](https://github.com/skyportal/skyportal/blob/b93d10f46ebcedcf13003381b9bbb1e077b53e3c/skyportal/handlers/api/source.py#L653-L657).
Currently, that query produces something like this:
```
SELECT classifications.id,
       classifications.created_at,
       classifications.modified,
       classifications.classification,
       classifications.taxonomy_id,
       classifications.probability,
       classifications.author_id,
       classifications.author_name,
       classifications.obj_id
FROM   classifications
JOIN
       (
              SELECT classifications_1.id  AS id
              FROM   classifications       AS classifications_1
              JOIN   group_classifications AS group_classifications_1
              ON     classifications_1.id = group_classifications_1.classification_id
              JOIN   groups
              ON     groups.id = group_classifications_1.group_id
              JOIN   group_users AS group_users_1
              ON     groups.id = group_users_1.group_id
              JOIN   users
              ON     users.id = group_users_1.user_id
              WHERE  users.id = %(id_1)s) AS anon_1
ON     anon_1.id = classifications.id
JOIN
       (
              SELECT classifications_2.id AS id
              FROM   classifications      AS classifications_2
              JOIN   taxonomies
              ON     taxonomies.id = classifications_2.taxonomy_id
              JOIN
                     (
                            SELECT taxonomies.id AS id
                            FROM   taxonomies
                            JOIN   group_taxonomy AS group_taxonomy_1
                            ON     taxonomies.id = group_taxonomy_1.taxonomie_id
                            JOIN   groups
                            ON     groups.id = group_taxonomy_1.group_id
                            JOIN   group_users AS group_users_2
                            ON     groups.id = group_users_2.group_id
                            JOIN   users
                            ON     users.id = group_users_2.user_id
                            WHERE  users.id = %(id_2)s) AS anon_3
              ON     anon_3.id = taxonomies.id
              JOIN   objs
              ON     objs.id = classifications_2.obj_id) AS anon_2
ON     anon_2.id = classifications.id
WHERE  classifications.obj_id = %(obj_id_1)s
```
Note the `JOIN   objs ON     objs.id = classifications_2.obj_id) AS anon_2`. When this is `EXPLAIN`'ed , you'll note a sequential scan on the entire `Objs` table like this:
Screenshot 2021-04-14 155538

With the changes in this PR, you now get a query like this:
```
SELECT classifications.id,
       classifications.created_at,
       classifications.modified,
       classifications.classification,
       classifications.taxonomy_id,
       classifications.probability,
       classifications.author_id,
       classifications.author_name,
       classifications.obj_id
FROM   classifications
JOIN
       (
              SELECT classifications_1.id  AS id
              FROM   classifications       AS classifications_1
              JOIN   group_classifications AS group_classifications_1
              ON     classifications_1.id = group_classifications_1.classification_id
              JOIN   groups
              ON     groups.id = group_classifications_1.group_id
              JOIN   group_users AS group_users_1
              ON     groups.id = group_users_1.group_id
              JOIN   users
              ON     users.id = group_users_1.user_id
              WHERE  users.id = %(id_1)s) AS anon_1
ON     anon_1.id = classifications.id
JOIN
       (
              SELECT classifications_2.id AS id
              FROM   classifications      AS classifications_2
              JOIN   taxonomies
              ON     taxonomies.id = classifications_2.taxonomy_id
              JOIN
                     (
                            SELECT taxonomies.id AS id
                            FROM   taxonomies
                            JOIN   group_taxonomy AS group_taxonomy_1
                            ON     taxonomies.id = group_taxonomy_1.taxonomie_id
                            JOIN   groups
                            ON     groups.id = group_taxonomy_1.group_id
                            JOIN   group_users AS group_users_2
                            ON     groups.id = group_users_2.group_id
                            JOIN   users
                            ON     users.id = group_users_2.user_id
                            WHERE  users.id = %(id_2)s) AS anon_3
              ON     anon_3.id = taxonomies.id) AS anon_2
ON     anon_2.id = classifications.id
WHERE  classifications.obj_id = %(obj_id_1)s
```
On the same sample query on production that I posted the `EXPLAIN` screenshot above, this new query gives:
image
Note the total time of ~4ms instead of 1100ms

The changes in lines 587:591 is more of a preemptive optimization that I don't think has popped up yet. It applies the same sort of "ignore Public attributes" to `ComposedAccessControl` logic.

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -462,16 +470,16 @@ def query_accessible_rows(self, cls, user_or_token, columns=None):
...
             # get the rows of the target table that are accessible
             join_target = relationship.entity.class_
             logic = getattr(join_target, mode)
 
             if isinstance(logic, Public):
                 continue
 
+            # join the target table to the related table on the relationship
+            base = base.join(relationship.class_attribute)
+
```
Comment (kmshin1397) [inline app/models.py:482]: This is a relevant change (non-`black`)

File: app/models.py [hunk 2]
```diff
@@ -576,6 +584,11 @@ def query_accessible_rows(self, cls, user_or_token, columns=None):
         accessible_id_cols = []
 
         for access_control in self.access_controls:
+
+            # Just ignore public ACLs
+            if isinstance(access_control, Public):
+                continue
+
```
Comment (kmshin1397) [inline app/models.py:591]: This is a relevant change (non-`black`)

**Review by acrellin — APPROVED** [review]
Excellent catch!

**Review by dmitryduev — APPROVED** [review]
This is an excellent optimization! Happy to merge as is. @dannygoldstein do you want to take a look?

File: app/models.py (hunk 1 above)
Comment (dannygoldstein) [reply app/models.py:482]: Ah, excellent catch -

**Review by dannygoldstein — APPROVED** [review]

**Pull request merged**